File: stages-backend/src/main/java/com/gestion/stages_backend/conventions/application/service/ConventionService.java
package com.gestion.stages_backend.conventions.application.service;

import com.gestion.stages_backend.candidatures.domain.model.Candidature;
import com.gestion.stages_backend.candidatures.infrastructure.persistence.JpaCandidatureRepository;
import com.gestion.stages_backend.conventions.application.dto.ConventionRequest;
import com.gestion.stages_backend.conventions.application.dto.ConventionResponse;
import com.gestion.stages_backend.conventions.application.mapper.ConventionMapper;
import com.gestion.stages_backend.conventions.domain.model.Convention;
import com.gestion.stages_backend.conventions.domain.model.StatutConvention;
import com.gestion.stages_backend.conventions.infrastructure.persistence.JpaConventionRepository;
import com.gestion.stages_backend.identity.domain.model.User;
import com.gestion.stages_backend.identity.infrastructure.persistence.JpaUserRepository;
import com.gestion.stages_backend.shared.application.dto.PageResponse;
import com.gestion.stages_backend.shared.infrastructure.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;

@Service
@RequiredArgsConstructor
public class ConventionService {

    private final JpaConventionRepository conventionRepository;
    private final JpaCandidatureRepository candidatureRepository;
    private final JpaUserRepository userRepository;
    private final ConventionMapper conventionMapper;

    @Transactional
    public ConventionResponse creer(ConventionRequest request) {
        Candidature candidature = candidatureRepository.findById(request.getCandidatureId())
                .orElseThrow(() -> new ResourceNotFoundException("Candidature", request.getCandidatureId()));
        User enseignant = userRepository.findById(request.getEnseignantId())
                .orElseThrow(() -> new ResourceNotFoundException("Enseignant", request.getEnseignantId()));

        Convention convention = Convention.builder()
                .candidature(candidature)
                .enseignant(enseignant)
                .dateDebutStage(request.getDateDebutStage())
                .dateFinStage(request.getDateFinStage())
                .build();

        return conventionMapper.toResponse(conventionRepository.save(convention));
    }

    @Transactional
    public ConventionResponse validerEnseignant(Long id, String commentaire) {
        Convention convention = conventionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Convention", id));
        convention.setStatut(StatutConvention.VALIDEE_ENSEIGNANT);
        convention.setCommentaireEnseignant(commentaire);
        convention.setDateValidationEnseignant(LocalDateTime.now());
        return conventionMapper.toResponse(conventionRepository.save(convention));
    }

    @Transactional
    public ConventionResponse approuverAdmin(Long id, String commentaire) {
        Convention convention = conventionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Convention", id));
        convention.setStatut(StatutConvention.APPROUVEE_ADMIN);
        convention.setCommentaireAdmin(commentaire);
        convention.setDateApprouveAdmin(LocalDateTime.now());
        return conventionMapper.toResponse(conventionRepository.save(convention));
    }

    @Transactional
    public ConventionResponse rejeter(Long id, String commentaire) {
        Convention convention = conventionRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Convention", id));
        convention.setStatut(StatutConvention.REJETEE);
        convention.setCommentaireAdmin(commentaire);
        return conventionMapper.toResponse(conventionRepository.save(convention));
    }

    @Transactional(readOnly = true)
    public PageResponse<ConventionResponse> findByStatut(StatutConvention statut, Pageable pageable) {
        if (statut == null) {
            // Retourner toutes les conventions sans filtre
            return PageResponse.of(
                conventionRepository.findAll(pageable)
                    .map(conventionMapper::toResponse));
        }
        return PageResponse.of(
            conventionRepository.findByStatut(statut, pageable)
                .map(conventionMapper::toResponse));
    }
}

File: stages-frontend/src/app/core/services/convention.service.ts
import { Injectable } from "@angular/core";
import { HttpClient, HttpParams } from "@angular/common/http";
import { ApiResponse, PageResponse } from "../models/api.model";
import { ConventionResponse, StatutConvention } from "../models/convention.model";

@Injectable({ providedIn: "root" })
export class ConventionService {
  private api = "/api/conventions";
  constructor(private http: HttpClient) {}

  liste(statut?: StatutConvention, page = 0, size = 20) {
    let params = new HttpParams().set("page", page).set("size", size);
    if (statut) params = params.set("statut", statut);
    return this.http.get<ApiResponse<PageResponse<ConventionResponse>>>(this.api, { params });
  }

  validerEnseignant(id: number, commentaire?: string) {
    let params = new HttpParams();
    if (commentaire) params = params.set("commentaire", commentaire);
    return this.http.patch<ApiResponse<ConventionResponse>>(`${this.api}/valider/${id}/enseignant`, {}, { params });
  }

  approuverAdmin(id: number, commentaire?: string) {
    let params = new HttpParams();
    if (commentaire) params = params.set("commentaire", commentaire);
    return this.http.patch<ApiResponse<ConventionResponse>>(`${this.api}/valider/${id}/admin`, {}, { params });
  }

  rejeter(id: number, commentaire?: string) {
    let params = new HttpParams();
    if (commentaire) params = params.set("commentaire", commentaire);
    return this.http.patch<ApiResponse<ConventionResponse>>(`${this.api}/${id}/rejeter`, {}, { params });
  }
}
